core/domain/state_domain.py

```diff
@@ -1406,7 +1426,13 @@ def validate(self, exp_param_specs_dict, allow_null_interaction):
         self.recorded_voiceovers.validate(content_id_list)
 
     def get_training_data(self):
-        """Retrieves training data from the State domain object."""
+        """Retrieves training data from the State domain object.
+
+        Returns:
+            list(dict). Each dict contains the following keys:
+                    answer_group_index: int. The index of the answer group.
+                    answers: str. The answers from the interaction.
```

Comment:
I still don't think the type is correct. How did you determine it?

Reply:
I checked the state_domain_test file. I was struggling to trigger the State class to run a test but when I checked the `state_domain_test.py` I found that `answers` is a list containing strings. Found this in line 166-169. You can confirm this please.

Reply:
The test files aren't sufficient for this -- you'll need to actually test the running code on the dev server to be sure, as well as analyze the callsites.

Try figuring out (by searching the codebase) what methods/functions call this method, and figuring out what calls that, until you get to the controller layer and main.py (see the overview of the codebase [here](https://github.com/oppia/oppia/wiki/Overview-of-the-Oppia-codebase)). That should give you some idea of what action to take in the UI in order to trigger this method, and then you'll also have a better idea of what is being passed in. If you get stuck, post your analysis in full and we can help.

Reply:
I'm stuck! Let me drop the full analysis.

Reply:
I worked my way from `get_training_data()` in `state_domain.py` to LIne 47, `handle_trainable_states` in `classifier_service.py`. From there I got to the private function `_create_exploration` on line 615 in `exp_services.py`. This private function is then used in `save_new_exploration` on line 638 in `exp_services.py`. Line 354 of `creator_dashboard.py`. Line 215 calls this in `main.py`. THat's how I discovered what method/functions are being triggered.

Reply:
After trying to edit some of the files, I got a null value. I can also drop full analysis of how I arrived at null for get_training_data() if needed.

Reply:
@seanlip

Reply:
That's great. So it seems like the part you need is to figure out how to continue working onwards from main.py?

main.py is a list of routes -- it connects URLs to handlers. Look at feconf.py in the root directory to find the URL that relates to the handler. Then visit that URL on your local dev server (or, do the action on the relevant page so that it triggers a request to the URL) in order to trigger the backend handler and all the functions it calls. That way you can log the type in the backend and inspect it in your server logs, as well as see the code in action all the way from the handler endpoint to the file you're changing.

If you're not sure where to go in the frontend, try searching for the URL (or the first part of the URL) in the frontend code.

Also, feel free to let us know if you get stuck with any of this and we'd be happy to help (but please provide detailed information about what you've managed to do so far)!

> I can also drop full analysis of how I arrived at null for get_training_data() if needed.

As a general note, please always do this (i.e. err on adding more analysis). It makes it easier for the reviewer to answer your questions without the delay of an additional back-and-forth cycle.

Reply:
Thanks so much for this analysis. I'd begin working with this.